### PR TITLE
issue/906: Added start controller

### PR DESCRIPTION
**Pull request by oliverfoster**
#906 

    /*example

        "_start": {
            "_isEnabled": true,
            "_startIds": [
                {
                    "_className": ".size-small.touch, .size-medium.touch",
                    "_id": "co-05",
                    "_skipIfComplete": false
                },
                {
                    "_className": ".size-small.no-touch, .size-medium.no-touch",
                    "_id": "co-10",
                    "_skipIfComplete": false
                }
            ],
            "_id": "course",
            "_force": true
        },
    */

### CODE REVIEW

**Comment by brian-learningpool** [comment]
This is a nice feature for core, I was just looking at the extension version of this the other day.  

It would be nice to get some documentation on this somewhere, aside from the JSON example in the code, as the question of how to do this is quite commonly asked in the forums.

**Comment by taylortom** [comment]
+1, with the comments removed (these should stay in example.json I think)

**Comment by oliverfoster** [comment]
agreed

**Comment by lc-thomasberger** [comment]
+1 (update documentation)

**Comment by tomgreenfield** [comment]
+1. Please also add `adapt:start` to the events [documentation](https://github.com/adaptlearning/adapt_framework/wiki/Core-Events#adapt-built-in-events).

**Comment by oliverfoster** [comment]
@moloko  wants to extend with sting pages that hide once completed... might be worth considering how that would fit in with this before we merge

**Comment by moloko** [comment]
Have a `_skipIfCompleted` property that can be added alongside any `_id` property?

**Comment by oliverfoster** [comment]
Yup, then you could line them up in the screenSize array.. perhaps i should rename screensize?

**Comment by moloko** [comment]
yeah, maybe `_startIds` instead?

**Comment by oliverfoster** [comment]
@moloko  better?

**Comment by moloko** [comment]
@oliverfoster looks good to me, I may actually give it a go on the project I'm working on at the moment

File: src/core/js/startController.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                    } catch(e) {
+                        console.log("startController: cannot find id", item._id);
+                        continue;
+                    }
+                    
+                    if (skipIfComplete) {
+                        if (model.get("_isComplete")) continue;
+                    }
+
+                    if (className || $("html").is(className)) {
```
Comment (moloko) [inline src/core/js/startController.js]: shouldn't this be 
```
if (!className || $("html").is(className)) {
```
As it is, if you don't specify `_className` it won't select that startid
?
At the moment, if I don't specify a className is

Comment (oliverfoster) [inline src/core/js/startController.js]: Yup, go for it!

Comment (moloko) [inline src/core/js/startController.js]: If I make that change it seems to be working for not specifying `_className`; I've run out of time to test it with that in place though

Comment (oliverfoster) [inline src/core/js/startController.js]: Yea, I can see any other reason for my hands writing that if+or unless it was meant to have a !

**Comment by moloko** [comment]
+1

**Comment by taylortom** [comment]
+1

**Comment by moloko** [comment]
have updated the [events documentation](adaptlearning/adapt_framework/wiki/Core-Events), plse check what I've put is accurate

**Comment by oliverfoster** [comment]
yup, perfect

**Comment by oliverfoster** [comment]
do we have a section for framework configuration?

**Comment by moloko** [comment]
I'm not sure we do... sounds like a job for @chucklorenz ;-)

also presumably need some way of adding this to the AT since the framework itself doesn't have a properties.schema... is that something you can pick up @taylortom or @brian-learningpool ?

**Comment by tomgreenfield** [comment]
As discussed, `_skipIfComplete` doesn't seem to be working well with Spoor.

**Comment by oliverfoster** [comment]
nope, it's because the completion handlers are now async, spoor restores, start does it's thing, the completion handlers call back and then the contentObject rolls over to complete. nahh... interesting one

**Comment by oliverfoster** [comment]
that's a rather large change, but it basically waits until all the completion cascades and callbacks have fired before letting Adapt initialize. Could prove to be more handy than in this situation only. We should be able to use Adapt.deferUntilCompletionChecked anywhere else it may be needed.

**Pull request merged**